### ISSUE TITLE
tidb's float Precision is much big than mysql-5.7

### ISSUE THREAD

**Issue by marsishandsome** [pingcap/tidb#10587]
## Bug Report

1. What did you do?

run in tidb, result is `1e-45`
![image](https://user-images.githubusercontent.com/5574887/58298546-7fc88380-7e0e-11e9-92a0-1c47a6adab30.png)

run same command in mysql-5.7, result is ` 1.4013e-45`
![image](https://user-images.githubusercontent.com/5574887/58298548-82c37400-7e0e-11e9-9485-42837eed5c87.png)

2. What did you expect to see?
tidb's result should be same as mysql or not so different from mysql

3. What did you see instead?
tidb's result is `1e-45`
mysql's result is `1.4013e-45`

if we use java jdbc to read from tidb, we will get the following error:
```
'1.0E-45' in column '1' is outside valid range for the datatype FLOAT.
com.mysql.jdbc.exceptions.jdbc4.MySQLDataException: '1.0E-45' in column '1' is outside valid range for the datatype FLOAT.
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at com.mysql.jdbc.Util.handleNewInstance(Util.java:425)
	at com.mysql.jdbc.Util.getInstance(Util.java:408)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:926)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:897)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:886)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:860)
	at com.mysql.jdbc.ResultSetImpl.throwRangeException(ResultSetImpl.java:7058)
	at com.mysql.jdbc.ResultSetImpl.getFloatFromString(ResultSetImpl.java:2429)
	at com.mysql.jdbc.ResultSetImpl.getFloat(ResultSetImpl.java:2398)
	at com.mysql.jdbc.ResultSetImpl.getObject(ResultSetImpl.java:4510)
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

| Release Version: v3.0.0-beta.1-12-g5c7e70e7f
Git Commit Hash: 5c7e70e7f50e720dc9ae12e0c4518ce027ceb869
Git Branch: master
UTC Build Time: 2019-03-28 05:25:57
GoVersion: go version go1.12 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false |

## SIG slack channel

 [#sig-exec](https://slack.tidb.io/invite?team=tidb-community&channel=sig-exec&ref=high-performance)

## Score

300

## Mentor

- @lzmhhh123

**Comment by XuHuaiyu** [comment]
@marsishandsome 
Will JDBC raise the error when it tries to read the value from MySQL?

**Comment by marsishandsome** [comment]
> @marsishandsome
> Will JDBC raise the error when it tries to read the value from MySQL?

no. everything is right when using MySQL-5.7.

because `1.4013e-45` is a valid value for mysql `FLOAT` type, but `1e-45` is not.

**Comment by XuHuaiyu** [comment]
> because 1.4013e-45 is a valid value for mysql FLOAT type, but 1e-45 is not.
I do not understand what does this mean?
I tested with MySQL 5.7.

``` sql
```

**Comment by nullnotnil** [comment]
Confirming this still exists against master:
```sql
tidb> DROP TABLE IF EXISTS t;
Query OK, 0 rows affected (0.21 sec)

tidb> CREATE TABLE t (c float);
Query OK, 0 rows affected (0.09 sec)

tidb> INSERT INTO t VALUES (1.4e-45);
Query OK, 1 row affected (0.03 sec)

tidb> SELECT * FROM t;
+-------+
| c     |
+-------+
| 1e-45 |
+-------+
1 row in set (0.00 sec)

tidb> select tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v4.0.0-beta.2-745-g2b0b34b88
Edition: Community
Git Commit Hash: 2b0b34b88e43ad20f4e5ab1a0b5daf7ae6ff6046
Git Branch: master
UTC Build Time: 2020-07-09 10:12:32
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.00 sec)
```

**Comment by wjhuang2016** [comment]
Confirm that this bug had been fixed in master, but not in v4.0.8.

``` sql
mysql> select * from t;
+------------+
| c          |
+------------+
| 1.4013e-45 |
+------------+
1 row in set (0.00 sec)

mysql> select tidb_version();
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                            |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v4.0.0-beta.2-1735-g8ec67d63b
Edition: Community
Git Commit Hash: 8ec67d63bbcaac5a4da540da1aaf88118c3ca15e
Git Branch: master
UTC Build Time: 2020-12-08 06:35:28
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

**Comment by wjhuang2016** [comment]
Fixed by PR https://github.com/pingcap/tidb/pull/20657
